File: itmo-2s-scala-hw3/src/main/scala/range/Range.scala
package range

trait Range {
  def length: Int

  def intersect(other: Range): Range

  def union(other: Range): Option[Range]

  def isEmpty: Boolean

  def contains(i: Int): Boolean

  def contains(other: Range): Boolean

  def isIntersect(other: Range): Boolean

  def toList: List[Int]

  def minimum: Option[Int]

  def maximum: Option[Int]
}

object Range {
  def empty: Range = EmptyRange

  def fromInterval(min: Int, max: Int): Range = new ClosedRange(min, max)
}

File: itmo-2s-golang-hw3/pkg/range/range.go
package rangeI

type RangeInt interface {
	Length() int
	Intersect(other RangeInt) RangeInt
	Union(other RangeInt) (RangeInt, bool)
	IsEmpty() bool
	ContainsInt(i int) bool
	ContainsRange(other RangeInt) bool
	IsIntersect(other RangeInt) bool
	ToSlice() []int
	Minimum() (int, bool)
	Maximum() (int, bool)
	ToString() string
}

type RangeIntImpl struct {
}

func NewRangeInt(a, b int) RangeInt {
	if a > b {
		return EmptyRange{}
	} else {
		return ClosedRange{a, b}
	}
}
